File: src/main/java/me/mrscopes/beans/events/ChatListener.java
package me.mrscopes.beans.events;

import io.papermc.paper.event.player.AsyncChatEvent;
import me.mrscopes.beans.Beans;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.TextComponent;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.inventory.ItemStack;

public class ChatListener implements Listener {
    @EventHandler
    public void onChat(AsyncChatEvent event) {
        event.setCancelled(true);

        Component message = event.message();

        if (message.toString().contains("[item]")) {
            ItemStack item = event.getPlayer().getInventory().getItemInMainHand();
            if (!item.getType().equals(Material.AIR)) {
                Component name = item.displayName().hoverEvent(item.asHoverEvent());

                TextComponent component = Component.text()
                        .append(Component.text(item.getAmount() > 1 ? item.getAmount() + "x " : "", item.displayName().color()))
                        .append(name)
                        .build();

                message = message.replaceText(msg -> msg.once().match("\\[item\\]").replacement(component));
            }
        }

        Bukkit.broadcast(MiniMessage.miniMessage().deserialize("<gray><player><white>: <message>",
                Placeholder.component("player", Component.text(event.getPlayer().getName())),
                Placeholder.component("message", message)));

        Component discordMessage = message.replaceText(builder -> builder.match("@everyone").replacement("at everyone"));
        String messageContent = PlainTextComponentSerializer.plainText().serialize(discordMessage);
        Beans.getDiscord().getServerChat().sendMessage(String.format("%s: %s", event.getPlayer().getName(), messageContent)).queue();

    }
}

File: src/main/kotlin/me/mrscopes/beans/events/ChatListener.kt
package me.mrscopes.beans.events

import io.papermc.paper.event.player.AsyncChatEvent
import me.mrscopes.beans.Beans
import me.mrscopes.beans.color
import me.mrscopes.beans.plainText
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.TextReplacementConfig
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import java.util.UUID
import kotlin.collections.HashMap

class ChatListener : Listener {
    val antispam: HashMap<UUID, Long> = HashMap()

    @EventHandler
    fun onChat(event: AsyncChatEvent) {
        event.isCancelled = true

        val player = event.player

        if (player.hasPermission("beans.chat.spam")) {
            val uuid = event.player.uniqueId
            val now = System.currentTimeMillis()
            if (antispam.containsKey(uuid)) {
                val lastMessageTime = now - antispam[uuid]!!
                if (lastMessageTime < 2000) {
                    val remaining = ((2000 - lastMessageTime)).toString().split("")
                    var time = "${remaining[1]}.${remaining[2]}${remaining[3]}"
                    if (remaining.size < 6) {
                        time = "0.${remaining[1]}${remaining[2]}"
                    }
                    player.sendMessage("&cWait $time seconds to chat again.".color())
                    return
                }
            }

            antispam[uuid] = now
        }

        var message = event.message()
        var plainText = message.plainText()
        if (player.hasPermission("beans.chat.color"))
            message = plainText.color()

        if (plainText.contains("[item]")) {
            val item = player.inventory.itemInMainHand
            if (item.type != Material.AIR) {
                val name = item.displayName().hoverEvent(item.asHoverEvent())
                val component = Component.text()
                        .append(Component.text(if (item.amount > 1) item.amount.toString() + "x " else "", item.displayName().color()))
                        .append(name)
                        .build()
                message = message.replaceText { msg: TextReplacementConfig.Builder -> msg.once().match("\\[item]").replacement(component) }
            }
        }

        Bukkit.broadcast("&7${event.player.name}&f: $message".color())

        val discordMessage = message.replaceText { builder: TextReplacementConfig.Builder -> builder.match("@everyone").match("<@538205671712358450>").replacement("at everyone") }
        Beans.discord.serverChat.sendMessage("${event.player.name}: ${discordMessage.plainText()}").queue()
    }
}
